File: frontend/src/api/exam.ts
import request from '@/utils/request'

// 题目类型枚举
export enum QuestionType {
  SINGLE = 'single',
  MULTIPLE = 'multiple',
  JUDGE = 'judge',
  BLANK = 'blank',
  ESSAY = 'essay'
}

export interface QuestionOption {
  label: string; // A, B, C...
  text: string;
}

export interface QuestionItem {
  id?: number;
  type: QuestionType;
  content: string;
  options?: QuestionOption[];
  answer: string | string[]; // 单选是 "A", 多选是 ["A","B"]
  analysis?: string;
  difficulty: number; // 1, 2, 3
  tags?: string[];
  updated_at?: string;
}


// 定义分页返回结构
export interface QuestionPageResult {
  total: number;
  items: QuestionItem[];
}

// 1. 获取题目列表
export function getQuestions(params: { 
  type?: string; 
  difficulty?: number; 
  keyword?: string; 
  tag?: string;
  skip?: number;
  limit?: number; 
}) {
  return request<any, QuestionPageResult>({ // ✅ 泛型指定返回结构
    url: '/exam/questions',
    method: 'get',
    params
  })
}

// 2. 创建题目
export function createQuestion(data: QuestionItem) {
  return request({
    url: '/exam/questions',
    method: 'post',
    data
  })
}

// 3. 删除题目
export function deleteQuestion(id: number) {
  return request({
    url: `/exam/questions/${id}`,
    method: 'delete'
  })
}

// 4. 更新题目
export function updateQuestion(id: number, data: QuestionItem) {
  return request({
    url: `/exam/questions/${id}`,
    method: 'put',
    data
  })
}

// 批量导入题目
export function batchImportQuestions(file: File) {
  const formData = new FormData();
  formData.append('file', file);
  return request({
    url: '/exam/questions/batch',
    method: 'post',
    data: formData,
    headers: { 'Content-Type': 'multipart/form-data' },
    timeout: 300000 // 导入可能较慢，设置1分钟超时
  });
}


export interface QuestionStats {
  total: number;
  type_counts: Record<string, number>; // { single: 10, multiple: 5 }
  difficulty_counts: Record<string, number>; // { '1': 5, '2': 3 }
}

// 5. 获取题库统计
export function getQuestionStats() {
  return request({
    url: '/exam/questions/stats',
    method: 'get'
  })
}


// ✅ 新增：获取所有标签
export function getAllTags() {
  return request<any, string[]>({
    url: '/exam/questions/tags',
    method: 'get'
  })
}

// ✅ 新增：批量删除
export function batchDeleteQuestions(ids: number[]) {
  return request({
    url: '/exam/questions/batch-delete',
    method: 'post',
    data: ids
  })
}


// ===========================
// 试卷相关
// ===========================

export interface ExamItem {
  id: number;
  title: string;
  status: number; // 0草稿 1已发布 2已结束
  total_score: number;
  duration: number;
  question_count: number;
  created_at: string;
  class_names: string[];
  start_time?: string;
  end_time?: string;
}

// 1. 获取试卷列表
export function getExams() {
  return request<any, ExamItem[]>({
    url: '/exam/exams',
    method: 'get'
  })
}

// 2. 创建试卷 (手动/随机)
// mode: 1手动, 2随机
export function createExam(data: {
  title: string;
  mode: number;
  questions?: Array<{ question_id: number; score: number }>;
  random_config?: Array<{ type: string; difficulty: number; tag?: string; count: number; score: number }>;
}) {
  return request({
    url: '/exam/exams',
    method: 'post',
    data
  })
}

// ✅ 3. 删除试卷
export function deleteExam(id: number) {
  return request({
    url: `/exam/exams/${id}`,
    method: 'delete'
  })
}

// ✅ 获取试卷详情
export function getExamDetail(id: number) {
  return request({
    url: `/exam/exams/${id}`,
    method: 'get'
  })
}

// ✅ 更新试卷
export function updateExam(id: number, data: any) {
  return request({
    url: `/exam/exams/${id}`,
    method: 'put',
    data
  })
}


// ----- 学生端考试相关------
// 1. 获取学生可见的考试列表
export function getStudentExams() {
  return request<any, any[]>({
    url: '/exam/student/list',
    method: 'get'
  })
}

// 2. 开始/进入考试
export function enterExam(examId: number) {
  return request({
    url: `/exam/student/enter/${examId}`,
    method: 'post'
  })
}

// 3. 获取试卷内容 (题目)
export function getExamPaper(examId: number) {
  return request<any, any[]>({
    url: `/exam/student/paper/${examId}`,
    method: 'get'
  })
}

// 4. 提交试卷
export function submitExam(examId: number, data: { answers: any[], cheat_count: number }) {
  return request({
    url: `/exam/student/submit/${examId}`,
    method: 'post',
    data
  })
}

// 5. 保存考试进度（暂存答案）
export function saveExamProgress(examId: number, answers: Array<{question_id: number, answer_content: any}>) {
  return request({
    url: `/exam/student/save-progress/${examId}`,
    method: 'post',
    data: answers
  })
}

// 6. 获取暂存的进度
export function getExamProgress(examId: number) {
  return request({
    url: `/exam/student/progress/${examId}`,
    method: 'get'
  })
}



// ===========================
// 阅卷相关
// ===========================

// 1. 获取某场考试的学生成绩列表
export function getExamRecords(examId: number) {
  return request<any, any[]>({
    url: `/exam/exams/${examId}/records`,
    method: 'get'
  })
}

// 2. 获取单份试卷详情 (阅卷用)
export function getRecordDetail(recordId: number) {
  return request<any, any>({
    url: `/exam/records/${recordId}/detail`,
    method: 'get'
  })
}

// 3. 提交评分
export function submitGrade(recordId: number, items: Array<{ question_id: number; score: number; feedback?: string }>) {
  return request({
    url: `/exam/records/${recordId}/grade`,
    method: 'post',
    data: { items }
  })
}

// 导出成绩单
export function exportExamGrades(examId: number) {
  return request({
    url: `/exam/exams/${examId}/export`,
    method: 'get',
    responseType: 'blob' // ✅ 关键：必须指定响应类型为二进制流
  })
}

// 实时查询库存
export function checkQuestionStock(data: { type: string; difficulty: number; tag?: string }) {
  return request<any, { count: number }>({
    url: '/exam/questions/check-stock',
    method: 'post',
    data
  })
}

// 预览随机组卷结果
export function previewRandomGeneration(strategies: any[]) {
  return request<any, any[]>({
    url: '/exam/exams/preview-random',
    method: 'post',
    data: strategies
  })
}

// 获取学生考试结果详情
export function getStudentResultDetail(examId: number) {
  return request<any, any>({
    url: `/exam/student/result/${examId}`,
    method: 'get'
  })
}

File: backend/app/models/exam.py
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, Text, JSON, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.base_class import Base

# ==========================================
# 1. 题库表 (Question)
# ==========================================
class Question(Base):
    __tablename__ = "questions"

    id = Column(Integer, primary_key=True, index=True)
    teacher_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # 题型: single(单选), multiple(多选), judge(判断), blank(填空), essay(简答)
    type = Column(String(20), nullable=False, index=True)
    
    # 题干 (富文本)
    content = Column(Text, nullable=False)
    
    # 选项 (仅选择题使用) -> 存 JSON: [{"label":"A", "text":"..."}, {"label":"B", "text":"..."}]
    options = Column(JSON, nullable=True)
    
    # 标准答案 -> 单选存字符串 "A"，多选存数组 ["A","B"]，填空存关键词数组
    answer = Column(JSON, nullable=False)
    
    # 解析
    analysis = Column(Text, nullable=True)
    
    # 难度: 1=易, 2=中, 3=难
    difficulty = Column(Integer, default=1)
    
    # 标签/知识点 -> 存 JSON: ["第一章", "计算机基础"]
    tags = Column(JSON, nullable=True)
    
    created_at = Column(DateTime, default=func.now())
    updated_at = Column(DateTime, default=func.now(), onupdate=func.now())

    # 关联
    teacher = relationship("User", backref="questions")


# ==========================================
# 2. 试卷主表 (Exam)
# ==========================================
class Exam(Base):
    __tablename__ = "exams"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(200), nullable=False)
    teacher_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # 考试配置
    start_time = Column(DateTime, nullable=True) # 允许开始时间
    end_time = Column(DateTime, nullable=True)   # 截止时间
    duration = Column(Integer, default=60)       # 考试时长(分钟)
    pass_score = Column(Integer, default=60)     # 及格分
    total_score = Column(Integer, default=100)   # 总分 (由题目累加或随机生成指定)
    
    # 状态: 0=草稿, 1=已发布, 2=已结束
    status = Column(Integer, default=0)
    
    # 发布给哪些班级 (存班级ID数组: [1, 2])
    class_ids = Column(JSON, nullable=True)
    
    # 组卷模式: 1=手动, 2=随机
    mode = Column(Integer, default=1)
    
    # 随机组卷的策略配置 (仅 mode=2 时有效)
    # 格式: [{"type":"single", "count":10, "score":2, "tags":["物流"], "diff":1}, ...]
    random_config = Column(JSON, nullable=True)

    created_at = Column(DateTime, default=func.now())
    updated_at = Column(DateTime, default=func.now(), onupdate=func.now())

    # 关联
    questions = relationship("ExamQuestion", back_populates="exam", cascade="all, delete-orphan") # 稍后定义

# ==========================================
# 3. 试卷-题目关联表 (ExamQuestion)
# ==========================================
class ExamQuestion(Base):
    __tablename__ = "exam_questions"

    id = Column(Integer, primary_key=True, index=True)
    exam_id = Column(Integer, ForeignKey("exams.id"), nullable=False)
    question_id = Column(Integer, ForeignKey("questions.id"), nullable=False)
    
    # 该题目在本试卷中的分值 (覆盖题库默认分值)
    score = Column(Integer, default=2)
    
    # 题目排序 (手动组卷用)
    sort_order = Column(Integer, default=0)

    # 关联
    exam = relationship("Exam", back_populates="questions")
    question = relationship("Question")


# ==========================================
# 4. 考试记录表 (ExamRecord)
# ==========================================
class ExamRecord(Base):
    __tablename__ = "exam_records"

    id = Column(Integer, primary_key=True, index=True)
    exam_id = Column(Integer, ForeignKey("exams.id"), nullable=False)
    student_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    # 时间记录
    start_time = Column(DateTime, default=func.now()) # 开始答题时间
    submit_time = Column(DateTime, nullable=True)     # 交卷时间
    
    # 状态: 0=进行中, 1=已提交(待批改), 2=已批改(完成)
    status = Column(Integer, default=0)
    
    # 分数统计
    objective_score = Column(Integer, default=0) # 客观题得分(系统自动判分)
    subjective_score = Column(Integer, default=0) # 主观题得分(老师人工判分)
    total_score = Column(Integer, default=0)      # 总得分
    
    # 监考数据
    cheat_count = Column(Integer, default=0) # 切屏次数/作弊嫌疑记录
    
    created_at = Column(DateTime, default=func.now())
    updated_at = Column(DateTime, default=func.now(), onupdate=func.now())

    # 关联
    exam = relationship("Exam", backref="records")
    student = relationship("User", foreign_keys=[student_id])
    answers = relationship("ExamAnswer", back_populates="record", cascade="all, delete-orphan")

# ==========================================
# 5. 学生答题明细表 (ExamAnswer)
# ==========================================
class ExamAnswer(Base):
    __tablename__ = "exam_answers"

    id = Column(Integer, primary_key=True, index=True)
    record_id = Column(Integer, ForeignKey("exam_records.id"), nullable=False)
    question_id = Column(Integer, ForeignKey("questions.id"), nullable=False)
    
    # 学生填写的答案 (单选:"A", 多选:["A","B"], 填空/简答: "文本内容")
    answer_content = Column(JSON, nullable=True)
    
    # 判分结果
    is_correct = Column(Boolean, nullable=True) # 是否正确 (主观题初始为None)
    score = Column(Integer, default=0)           # 该题实得分
    
    # 老师评价
    teacher_feedback = Column(Text, nullable=True)

    # 关联
    record = relationship("ExamRecord", back_populates="answers")
    question = relationship("Question")